notifications/manager.go

```diff
@@ -0,0 +1,193 @@
+		chan *swapserverrpc.ServerReservationNotification, //nolint:lll
+	)
+	sub := subscriber{
+		subCtx:   ctx,
+		recvChan: notifChan,
+	}
+
+	m.Lock()
+	m.subscribers[NotificationTypeReservation] = append(
+		m.subscribers[NotificationTypeReservation],
+		sub,
+	)
+	m.Unlock()
+
+	return notifChan
+}
+
+// Run starts the notification manager.
+func (n *Manager) Run(ctx context.Context, readyChan chan<- struct{}) error {
+	// In order to create a valid l402 we first are going to call
+	// the FetchL402 method. As a client might not have outbound capacity
+	// yet, we'll retry until we get a valid response.
+	if !n.hasL402 {
+		n.fetchL402(ctx)
+	}
+
+	var closedChan bool
+
+	// Start the notification runloop.
+	for {
+		// Return if the context has been canceled.
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		callCtx, cancel := context.WithCancel(ctx)
+		notifStream, err := n.cfg.Client.SubscribeNotifications(
+			callCtx, &swapserverrpc.SubscribeNotificationsRequest{},
+		)
+		if err != nil {
+			cancel()
+			log.Errorf("Error subscribing to notifications: %v", err)
+			continue
+		}
+		if !closedChan {
+			close(readyChan)
+		}
+
+		log.Debugf("Successfully subscribed to server notifications")
+
+		for {
+			notification, err := notifStream.Recv()
+			if err == nil && notification != nil {
+				log.Debugf("Received notification: %v", notification)
+				n.handleNotification(notification)
+				continue
+			}
+			log.Errorf("Error receiving "+
+				"notification: %v", err)
+
+			cancel()
+			// Wait for a bit before reconnecting.
+			<-time.After(time.Second * 10)
+			break
+		}
+	}
+}
+
+// fetchL402 fetches the L402 from the server. This method will keep on
+// retrying until it gets a valid response.
+func (m *Manager) fetchL402(ctx context.Context) {
+	// Add a 0 timer so that we initially fetch the L402 immediately.
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-timer.C:
+			err := m.cfg.FetchL402(ctx)
+			if err != nil {
+				log.Warnf("Error fetching L402: %v", err)
+				timer.Reset(time.Second * 10)
+				continue
+			}
+			m.hasL402 = true
+
+			return
+		}
+	}
+}
+
+// handleNotification handles an incoming notification from the server,
+// forwarding it to the appropriate subscribers.
+func (n *Manager) handleNotification(notification *swapserverrpc.
+	SubscribeNotificationsResponse) {
+
+	switch notification.Notification.(type) {
+	case *swapserverrpc.SubscribeNotificationsResponse_ReservationNotification:
+		// We'll forward the reservation notification to all subscribers.
+		// Cleaning up any subscribers that have been canceled.
+		newSubs := []subscriber{}
+		reservationNtfn := notification.GetReservationNotification()
+		for _, sub := range n.subscribers[NotificationTypeReservation] {
+			recvChan := sub.recvChan.(chan *swapserverrpc.
+				ServerReservationNotification)
+
+			select {
+			case <-sub.subCtx.Done():
+				close(recvChan)
+				continue
```

Comment:
Should it be `break` here? Otherwise we can close recvChan multiple times (panic).

Reply:
You are right! I confused sub.subCtx with ctx :)